### PR TITLE
[Bridge/Twig] Let getFlashes starts the session

### PR DESCRIPTION
**Pull request by MatTheCat**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    |no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25073 
| License       | MIT

### CODE REVIEW

File: src/Symfony/Bridge/Twig/Tests/AppVariableTest.php [hunk 1]
```diff
@@ -258,7 +253,7 @@ protected function setTokenStorage($user)
         $token->method('getUser')->willReturn($user);
     }
 
-    private function setFlashMessages()
+    private function setFlashMessages($startSession = true)
```
Comment (sroze) [inline src/Symfony/Bridge/Twig/Tests/AppVariableTest.php]: I think you should rename to `sessionHasStarted`

**Comment by sroze** [comment]
@MatTheCat can you change the base of the pull-request to be the 3.4 branch on the UI?

File: src/Symfony/Bridge/Twig/Tests/AppVariableTest.php (hunk 1 above)
Comment (MatTheCat) [reply src/Symfony/Bridge/Twig/Tests/AppVariableTest.php]: Done

**Comment by MatTheCat** [comment]
@sroze done

**Comment by javiereguiluz** [comment]
@MatTheCat I'm shocked about this proposal 😱  We created `getFlash()` primarily to avoid starting the session when asking for flash messages. This behavior was reported multiple times by users ... and I personally consider this a Symfony bug (although some disagree and this was never "fixed" or changed).

**Comment by MatTheCat** [comment]
I thought the response was to check for `app.request.hasPreviousSession` as mentioned in http://symfony.com/doc/current/session/avoid_session_start.html? Following this advice it really is counter-intuituve to start the session from Twig.

**Comment by MatTheCat** [comment]
Also I find very weird we focus DX for people who read read session data while not wanting any session?

**Comment by javiereguiluz** [comment]
@MatTheCat I don't know the internal details about this (@nicolas-grekas and @sroze will help us here) but if I ask Symfony whether there are flash messages or not ... I don't want the session to be started if it wasn't already started ... because that hurts app performance. I'm just asking if there are flash messages; I'm not getting them. That's why I think the session should not start. Thanks!

**Comment by MatTheCat** [comment]
@javiereguiluz but by returning an empty array how can the presence of flashes be checked?

**Comment by javiereguiluz** [comment]
If flash messages are stored in the session ... and the session has not even been started ... the answer to "are there flash messages?" should be no ... and that's why we return an empty array, to say that there are no flash messages.

**Comment by nicolas-grekas** [comment]
@javiereguiluz AFAIK, the session is "started" when the backend has been read. I don't think we can know if any flash are set if we don't read the backend.
The current logic is broken, because it relies on the fact that a previous session entry (not flash) has been read, so that *then* the session has started, an flash are fetched already.
The performance issue of opening an empty session is gone with #24523, so the workaround can be removed, and the logic be fixed now. It should really.

**Comment by MatTheCat** [comment]
Buy you can't just say a box is empty because you don't want to open it?!

**Comment by javiereguiluz** [comment]
@MatTheCat following your analogy, I'm saying: if I see there's no box, then I can say there's nothing inside the box. And you are saying (that's at least what I understand): if I see there's no box, then I'm going to create a new box an open it to see if there's something inside it.

But @nicolas-grekas said this is OK, so it's clear that I'm not understanding this. So ignore my comments :) Thanks!

**Comment by MatTheCat** [comment]
Yeah okay I didn't understand the relation with #24523

**Comment by nicolas-grekas** [comment]
@javiereguiluz I think the misunderstanding comes from what "session is started" means. When there is no session cookie, no "box" is created anymore. But when there is a session cookie, we must ask the backend for a corresponding "box" there. "Starting the session" means actually this: has the potentially existing box been opened? If not, telling there are no flashes there is just wrong.

**Comment by MatTheCat** [comment]
@javiereguiluz that's it! I don't know about performance but I would prefer to create an empty box than always saying it's empty.

**Comment by MatTheCat** [comment]
(My comment was "But you can't say a box is empty just because you don't want to open it?!")

**Comment by javiereguiluz** [comment]
Thanks Nico. That was what I wasn't understanding: "starting a session" and "creating a session" were the same to me ... but they are not! Thanks!

**Comment by nicolas-grekas** [comment]
@MatTheCat would you mind rebasing on latest 3.4? (you can even squash)
Tests did not run yet, that would fix that.

**Comment by MatTheCat** [comment]
@nicolas-grekas done

**Review by nicolas-grekas — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @MatTheCat.

**Comment by Vedmak** [comment]
> ... When there is no session cookie, no "box" is created anymore. ...

@nicolas-grekas If i have no session cookie and call app.flashes, new session and cookie is created. 
Is it correct behavior? 
I'm just not sure if i understood this discussion correctly. Also its quite misleading after reading https://symfony.com/blog/new-in-symfony-3-3-improved-flash-messages , i understand that this fix is for 3.4, but article is about 3.3, but it is one of top links when you  google for "symfony flashes"

**Comment by nicolas-grekas** [comment]
3.4 embeds 3.3 so that the article applies to it also. I can't tell from afar if it's legit for the session to be started or not. If you think it's a bug, please open a dedicated issue with a reproducer, i.e. a repository we could clone to run the issue locally.

**Comment by Vedmak** [comment]
Sorry for delayed answer.
But i'm not sure if it's a bug and if it is a bug, is it a documentation bug or code bug, it's what i'm trying to find out. Documentations about session starting for version [3.3](https://symfony.com/doc/3.3/session/avoid_session_start.html), [3.4](https://symfony.com/doc/3.4/session/avoid_session_start.html) and [current](https://symfony.com/doc/current/session/avoid_session_start.html) all tell that

> Even if the user is not logged in and even if you haven't created any flash messages, just calling the get() (or even has()) method of the flashBag will start a session.

But [blog post](messages) i mentioned previously, about improved flash messages states

> Lastly, the new helper fixes one of the most frustrating issues related to flash messages: checking for the existence of flash messages starts the session automatically. That's why you had to check first if the session already existed. This check is now done internally, so you no longer have to care about this and the session will never start automatically again

**Pull request merged**